concurrency/GetProductPromiseActivity.java

```diff
@@ -0,0 +1,287 @@
+        catalogInformation.setOfferSKU(input.getOfferSKU());
+
+        Product product = new Product();
+        product.setOfferListingInformation(offerListingInformation);
+        product.setCatalogInformation(catalogInformation);
+
+        SellerInformation sellerInformation = new SellerInformation();
+        sellerInformation.setMarketplaceID(input.getMarketplaceId().getEncryptedId());
+        sellerInformation.setMerchantID(input.getMerchantId().getEncryptedId());
+        sellerInformation.setMerchantCustomerID(sellerOfRecordForMerchant.getSORMerchant().getMerchantCustomerID());
+        sellerInformation.setSellerOfRecordID(sellerOfRecordForMerchant.getSORMerchant().getSOR().getLegalEntityID());
+
+        GetProductDeliveryInformationSummary.Request request = pdexAggregatorServiceClient
+                .newGetProductDeliveryInformationSummaryRequest(product, sellerInformation);
+        request.setMaximumDeliveryOptionCount(MAX_NUM_OF_DELIVERY_OPTIONS);
+        if (input.getDestination() != null) {
+            com.amazon.afsorderservice.Address inputDestination = input.getDestination();
+            Address address = new Address();
+            address.setCountryCode(inputDestination.getCountryCode());
+            address.setPostalCode(inputDestination.getPostalCode());
+            address.setStateOrRegion(inputDestination.getStateOrProvinceCode());
+            address.setCity(inputDestination.getCity());
+            address.setDistrictOrCounty(inputDestination.getDistrict());
+            request.setDestination(address);
+        }
+        return request;
+    }
+
+    /**
+     * This map is to store sellerOfOrderRecord result and avoid redundant duplicated call to PARIS.
+     * For now we're not sure how many merchants will there be, just assume the merchant count is
+     * not large and thus use this temporary way as a cache.
+     * <p>
+     * When merchant count get larger, will change to use standard caching technologies in
+     * https://w.amazon.com/index.php/CachingTeam
+     */
+    private static ConcurrentHashMap<Pair<String, String>, SORMerchantMP> sorMerchantMPMap = new ConcurrentHashMap<>();
+
+    /**
+     * @param marketplaceId      the encrypted marketplaceId
+     * @param merchantCustomerId the encrypted merchant customerId
+     * @return the seller of record response from PARIS
+     * @throws Exception when the call fails
+     */
+    protected SORMerchantMP getSellerOfRecordForMerchant(
+            final String marketplaceId,
+            final String merchantCustomerId) {
+        try {
+            Pair<String, String> keyPair = Pair.of(marketplaceId, merchantCustomerId);
+            SORMerchantMP sorMerchantMP = sorMerchantMPMap.get(keyPair);
+            if (null == sorMerchantMP) {
+                sorMerchantMP = parisService.getDefaultSORByMerchantMP(marketplaceId, merchantCustomerId);
+                sorMerchantMPMap.put(keyPair,sorMerchantMP);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **273, 274, and 276**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **278**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.